libbpf-tools/readahead.bpf.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "readahead.h"
+#include "bits.bpf.h"
+
+#define MAX_ENTRIES 10240
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, u8);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} in_readahead SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct page *);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} birth SEC(".maps");
+
+static struct hist hist;
+
+SEC("fentry/__do_page_cache_readahead")
+int BPF_PROG(fentry____do_page_cache_readahead)
+{
+	u32 pid = bpf_get_current_pid_tgid();
+
+	bpf_map_update_elem(&in_readahead, &pid, &pid, 0);
+	return 0;
+}
+
+SEC("fexit/__page_cache_alloc")
+int BPF_PROG(fexit____page_cahe_alloc, gfp_t gfp, struct page *ret)
+{
+	u32 pid = bpf_get_current_pid_tgid(), ts;
```

Comment:
ts has to be u64

Reply:
Oh so sorry, I changed `u64 id` to `u32 pid`, but forgot to check ts's type.

Reply:
Fixed, thanks!